File: adaptor/go/kafka.go
package microscope

import (
	"context"
	"time"

	"github.com/segmentio/kafka-go"
)

// KafkaWriter instruments a kafka-go writer with Redpanda topic signals.
type KafkaWriter struct {
	Writer *kafka.Writer
	Hub    *Hub
}

// NewKafkaWriter wraps writer with topic instrumentation.
func NewKafkaWriter(writer *kafka.Writer, hub *Hub) *KafkaWriter {
	return &KafkaWriter{Writer: writer, Hub: hub}
}

// WriteMessages writes messages and records one signal for the produced batch.
func (w *KafkaWriter) WriteMessages(ctx context.Context, messages ...kafka.Message) error {
	started := time.Now()
	err := w.Writer.WriteMessages(ctx, messages...)
	topic := w.Writer.Topic
	if topic == "" && len(messages) > 0 {
		topic = messages[0].Topic
	}
	var bytes int
	for _, message := range messages {
		bytes += len(message.Key) + len(message.Value)
	}
	content := map[string]any{
		"message_count": len(messages),
		"size_bytes":    bytes,
		"error":         errorString(err),
	}
	if w.Hub != nil && !w.Hub.RedactSensitive() {
		maxBytes := w.Hub.cfg.MaxBodyBytes
		if maxBytes <= 0 {
			maxBytes = 65536
		}
		payloads := make([]map[string]string, 0, len(messages))
		for _, message := range messages {
			payloads = append(payloads, map[string]string{
				"key":   TruncateBytes(message.Key, maxBytes),
				"value": TruncateBytes(message.Value, maxBytes),
			})
		}
		content["messages"] = payloads
	}
	w.Hub.RecordTopic(ctx, topic, "produce", time.Since(started), content)
	return err
}

// Close closes the wrapped writer.
func (w *KafkaWriter) Close() error {
	return w.Writer.Close()
}

// KafkaReader instruments kafka-go fetch and commit operations.
type KafkaReader struct {
	Reader *kafka.Reader
	Hub    *Hub
}

// NewKafkaReader wraps reader with topic instrumentation.
func NewKafkaReader(reader *kafka.Reader, hub *Hub) *KafkaReader {
	return &KafkaReader{Reader: reader, Hub: hub}
}

// FetchMessage fetches a message and records its topic, partition, and offset.
func (r *KafkaReader) FetchMessage(ctx context.Context) (kafka.Message, error) {
	started := time.Now()
	message, err := r.Reader.FetchMessage(ctx)
	topic := message.Topic
	if topic == "" {
		topic = r.Reader.Config().Topic
	}
	content := map[string]any{
		"partition":  message.Partition,
		"offset":     message.Offset,
		"size_bytes": len(message.Key) + len(message.Value),
		"error":      errorString(err),
	}
	if r.Hub != nil && !r.Hub.RedactSensitive() {
		maxBytes := r.Hub.cfg.MaxBodyBytes
		if maxBytes <= 0 {
			maxBytes = 65536
		}
		content["key"] = TruncateBytes(message.Key, maxBytes)
		content["value"] = TruncateBytes(message.Value, maxBytes)
	}
	r.Hub.RecordTopic(ctx, topic, "consume", time.Since(started), content)
	return message, err
}

// CommitMessages commits messages and records the acknowledgement operation.
func (r *KafkaReader) CommitMessages(ctx context.Context, messages ...kafka.Message) error {
	started := time.Now()
	err := r.Reader.CommitMessages(ctx, messages...)
	topic := r.Reader.Config().Topic
	if topic == "" && len(messages) > 0 {
		topic = messages[0].Topic
	}
	r.Hub.RecordTopic(ctx, topic, "commit", time.Since(started), map[string]any{
		"message_count": len(messages),
		"error":         errorString(err),
	})
	return err
}

// Close closes the wrapped reader.
func (r *KafkaReader) Close() error {
	return r.Reader.Close()
}

func errorString(err error) string {
	if err == nil {
		return ""
	}
	return err.Error()
}

File: adaptor/python/microscope/kafka.py
from __future__ import annotations

import time
from typing import Any

from microscope.hub import Hub
from microscope.sanitize import truncate_bytes


class KafkaWriter:
    def __init__(self, writer: Any, hub: Hub) -> None:
        self._writer = writer
        self._hub = hub

    def write_messages(self, *args, **kwargs) -> Any:
        messages = kwargs.get("messages")
        if messages is None and args:
            messages = args[0]
        messages = list(messages or [])
        started = time.monotonic()
        error = None
        try:
            return self._writer.write_messages(*args, **kwargs)
        except Exception as exc:
            error = exc
            raise
        finally:
            topic = getattr(self._writer, "topic", "") or ""
            if not topic and messages:
                topic = getattr(messages[0], "topic", "") or messages[0].get("topic", "")
            size_bytes = 0
            for message in messages:
                key = getattr(message, "key", None) or message.get("key", b"")
                value = getattr(message, "value", None) or message.get("value", b"")
                size_bytes += len(key or b"") + len(value or b"")
            content: dict[str, Any] = {
                "message_count": len(messages),
                "size_bytes": size_bytes,
                "error": str(error) if error else "",
            }
            if not self._hub.redact_sensitive():
                max_bytes = self._hub.config.max_body_bytes or 65536
                payloads = []
                for message in messages:
                    key = getattr(message, "key", None) or message.get("key", b"")
                    value = getattr(message, "value", None) or message.get("value", b"")
                    payloads.append(
                        {
                            "key": truncate_bytes(bytes(key or b""), max_bytes),
                            "value": truncate_bytes(bytes(value or b""), max_bytes),
                        }
                    )
                content["messages"] = payloads
            self._hub.record(
                "topic",
                {
                    "topic": topic,
                    "action": "produce",
                    "duration_ms": (time.monotonic() - started) * 1000,
                    **content,
                },
                tags=[f"topic:{topic}", "kafka:produce"],
            )

    def close(self) -> Any:
        return self._writer.close()


class KafkaReader:
    def __init__(self, reader: Any, hub: Hub) -> None:
        self._reader = reader
        self._hub = hub

    def fetch_message(self, *args, **kwargs) -> Any:
        started = time.monotonic()
        error = None
        message = None
        try:
            message = self._reader.fetch_message(*args, **kwargs)
            return message
        except Exception as exc:
            error = exc
            raise
        finally:
            topic = ""
            partition = 0
            offset = 0
            key = b""
            value = b""
            if message is not None:
                topic = getattr(message, "topic", "") or message.get("topic", "")
                partition = getattr(message, "partition", 0) or message.get("partition", 0)
                offset = getattr(message, "offset", 0) or message.get("offset", 0)
                key = getattr(message, "key", None) or message.get("key", b"")
                value = getattr(message, "value", None) or message.get("value", b"")
            if not topic:
                topic = getattr(self._reader, "topic", "") or getattr(self._reader, "_topic", "")
            content: dict[str, Any] = {
                "partition": partition,
                "offset": offset,
                "size_bytes": len(key or b"") + len(value or b""),
                "error": str(error) if error else "",
            }
            if not self._hub.redact_sensitive():
                max_bytes = self._hub.config.max_body_bytes or 65536
                content["key"] = truncate_bytes(bytes(key or b""), max_bytes)
                content["value"] = truncate_bytes(bytes(value or b""), max_bytes)
            self._hub.record(
                "topic",
                {
                    "topic": topic,
                    "action": "consume",
                    "duration_ms": (time.monotonic() - started) * 1000,
                    **content,
                },
                tags=[f"topic:{topic}", "kafka:consume"],
            )

    def commit_messages(self, *args, **kwargs) -> Any:
        messages = kwargs.get("messages")
        if messages is None and args:
            messages = args[0]
        messages = list(messages or [])
        started = time.monotonic()
        error = None
        try:
            return self._reader.commit_messages(*args, **kwargs)
        except Exception as exc:
            error = exc
            raise
        finally:
            topic = getattr(self._reader, "topic", "") or getattr(self._reader, "_topic", "")
            if not topic and messages:
                message = messages[0]
                topic = getattr(message, "topic", "") or message.get("topic", "")
            self._hub.record(
                "topic",
                {
                    "topic": topic,
                    "action": "commit",
                    "duration_ms": (time.monotonic() - started) * 1000,
                    "message_count": len(messages),
                    "error": str(error) if error else "",
                },
                tags=[f"topic:{topic}", "kafka:commit"],
            )

    def close(self) -> Any:
        return self._reader.close()
